src/daemon/rpmostreed-os.cxx

```diff
@@ -932,6 +934,127 @@ os_handle_finalize_deployment (RPMOSTreeOS *interface, GDBusMethodInvocation *in
   return TRUE;
 }
 
+static DnfContext *
+os_create_bare_dnf_context (GCancellable *cancellable, GError **error)
+{
+  g_autoptr (DnfContext) dnfctx = dnf_context_new ();
+  g_autoptr (DnfState) state = NULL;
+
+  /* Use a bare DNF context for the search */
+  dnf_context_set_repo_dir (dnfctx, "/etc/yum.repos.d");
+  dnf_context_set_cache_dir (dnfctx, RPMOSTREE_CORE_CACHEDIR RPMOSTREE_DIR_CACHE_REPOMD);
+  dnf_context_set_solv_dir (dnfctx, RPMOSTREE_CORE_CACHEDIR RPMOSTREE_DIR_CACHE_SOLV);
+  dnf_context_set_cache_age (dnfctx, G_MAXUINT);
+  dnf_context_set_enable_filelists (dnfctx, FALSE);
+
+  if (!dnf_context_setup (dnfctx, cancellable, error))
+    return NULL;
+
+  state = dnf_state_new ();
+  if (!dnf_context_setup_sack_with_flags (dnfctx, state, DNF_CONTEXT_SETUP_SACK_FLAG_SKIP_RPMDB,
+                                          error))
+    return NULL;
+
+  return (DnfContext *)g_steal_pointer (&dnfctx);
+}
+
+static void
+os_add_package_info_to_builder (DnfPackage *pkg, GVariantBuilder *builder, const gchar *key)
+{
+  GVariantDict pkg_dict;
+  g_variant_dict_init (&pkg_dict, NULL);
+  if (key != NULL)
+    g_variant_dict_insert (&pkg_dict, "key", "s", key);
+#define add_value_safe(_key, _value)                                                               \
+  g_variant_dict_insert (&pkg_dict, _key, "s", _value ? _value : "");
+  add_value_safe ("name", dnf_package_get_name (pkg));
+  g_variant_dict_insert (&pkg_dict, "epoch", "t", dnf_package_get_epoch (pkg));
+  add_value_safe ("version", dnf_package_get_version (pkg));
+  add_value_safe ("arch", dnf_package_get_arch (pkg));
+  add_value_safe ("nevra", dnf_package_get_nevra (pkg));
+  add_value_safe ("evr", dnf_package_get_evr (pkg));
+  add_value_safe ("summary", dnf_package_get_summary (pkg));
+  add_value_safe ("reponame", dnf_package_get_reponame (pkg));
+#undef add_value_safe
+  g_variant_builder_add_value (builder, g_variant_dict_end (&pkg_dict));
+}
+
+static gboolean
+os_handle_what_provides (RPMOSTreeOS *interface, GDBusMethodInvocation *invocation,
+                         const gchar *const *provides)
+{
+  GError *local_error = NULL;
+  g_autoptr (GCancellable) cancellable = NULL;
+  g_autoptr (DnfContext) dnfctx = NULL;
+
+  dnfctx = os_create_bare_dnf_context (cancellable, &local_error);
```

Comment:
This code isn't doing any locking, which means that e.g. two concurrent invocations of WhatProvides will happily try to fetch rpm-md in multiple threads in parallel, potentially racing on reading/writing the cached data.

What we should probably do is have the client explicitly call `RefreshMd` if desired before this - that way, we can reuse all the infrastructure we already have for things like progress reporting.

Then this code should:

- Grab some sort of "read-only" lock (today we have the concept of a transaction which acts a lot like a mutex, but we should probably convert that to a reader-writer lock conceptually).  
- Operate offline (via `rpmostree_context_set_dnf_caching (RPMOSTREE_CONTEXT_DNF_CACHE_FOREVER)`), assuming that callers that wanted to refresh metadata did it explicitly beforehand

Reply:
If I'm not mistaken, using `dnf_context_set_cache_age (dnfctx, G_MAXUINT);` and `DNF_CONTEXT_SETUP_SACK_FLAG_SKIP_RPMDB` makes sure there is no refresh of the rpmdb by these calls, is it not? The "bare" dbf context is meant to be a read-only variant of the context. I though the read/write locking of the DB is done on the libdnf side (the same as when you run two `dnf` commands from a terminal.

Anyway, if there is a real locking added to this, would it mean you also need to make it a real transaction, thus it does not block the D-Bus thread?

Reply:
You may be right that the libdnf level locking is sufficient.  But, honestly I'm a bit skeptical of that as I'm not aware of it being extensively tested.

In the end for now, I think gnome-software and similar tools should just not invoke things like `ListRepos` concurrently with doing upgrades.

Reply:
If the time shows a locking is needed I (or anybody else) can (I guess easily) turn those methods into regular transactions. I did think of the transactions at the beginning, but then I noticed the `GetDeploymentBootConfig`, which works similarly, thus I copied many of the bits from there. Being similar to this method the code is also simpler than with async transactions.